crates/sparrow-syntax/src/syntax/signature.rs

```diff
@@ -39,9 +44,44 @@ impl Signature {
             )
         }
 
+        debug_assert!(
+            type_parameters.is_empty(),
+            "passing type params unsupported"
+        );
+
+        // Contains the mapping of type constraints to their type variable name
+        // This just increments 'A' for each new constraint. This will need to change
+        // once we also support passing in type params through the signature definition.
+        let mut constraint_to_names: BTreeMap<TypeConstraint, u8> = BTreeMap::new();
+        let mut type_var = b'A';
+
+        // Check parameters for type constraints
+        for t in parameters.types() {
+            if let FenlType::Generic(constraint) = t.inner() {
+                if !constraint_to_names.contains_key(constraint) {
+                    constraint_to_names.insert(*constraint, type_var);
+                    type_var += 1;
+                }
+            }
+        }
+
+        // Check result for type constraint
+        if let FenlType::Generic(constraint) = result {
+            constraint_to_names.entry(constraint).or_insert(type_var);
```

Comment:
This assigns a type variable for a constraint that appears only in the result (which currently shouldn't happen), but it doesn't actually change the result to use that type variable, nor does it change the parameters.

I think as part of this we can change `FenlType` to include a `TypeVariable`, but not allow `TypeConstraint` to be used as `FenlType` (so generics in a `FenlType` are always a reference to a type variable).

Then, we need to change the parameters and result to reference the assigned type parameter.

That in turn will impact how we do inference.